{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d886b9e3-9dea-417f-9fc4-afd9dcd82da4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "61bb3bee-5324-4752-83ca-6697036ef039",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mutation_info_varscan(row):\n",
    "    try:\n",
    "        tumorAF = float(row[\"Otherinfo1\"].split(\"\\t\")[-1].split(\":\")[5].split(\"%\")[0])\n",
    "        normalAF = float(row[\"Otherinfo1\"].split(\"\\t\")[-2].split(\":\")[5].split(\"%\")[0])\n",
    "        tumorDepth = float(row[\"Otherinfo1\"].split(\"\\t\")[-1].split(\":\")[2])\n",
    "        normalDepth = float(row[\"Otherinfo1\"].split(\"\\t\")[-2].split(\":\")[2])\n",
    "        tumorRefReads = float(row[\"Otherinfo1\"].split(\"\\t\")[-1].split(\":\")[3])\n",
    "        normalRefReads = float(row[\"Otherinfo1\"].split(\"\\t\")[-2].split(\":\")[3])\n",
    "        tumorMutReads = float(row[\"Otherinfo1\"].split(\"\\t\")[-1].split(\":\")[4])\n",
    "        normalMutReads = float(row[\"Otherinfo1\"].split(\"\\t\")[-2].split(\":\")[4])\n",
    "        tumorGenotype = row[\"Otherinfo1\"].split(\"\\t\")[-1].split(\":\")[0]\n",
    "        normalGenotype = row[\"Otherinfo1\"].split(\"\\t\")[-2].split(\":\")[0]\n",
    "        #info_list = [tumorAF, normalAF, tumorDepth, normalDepth, tumorRefReads, normalRefReads, tumorMutReads, normalMutReads, tumorGenotype, normalGenotype]\n",
    "        #return info_list\n",
    "        return (tumorAF, normalAF, tumorDepth, normalDepth, tumorRefReads, normalRefReads, tumorMutReads, normalMutReads, tumorGenotype, normalGenotype)\n",
    "    except IndexError:\n",
    "        return None\n",
    "    \n",
    "def process_file(file_path, gene_list):\n",
    "    data_raw = pd.read_csv(file_path, low_memory=False)\n",
    "    data_raw['rating'] = -1\n",
    "    # 只保留exonic和splicing的变异\n",
    "    data_sub = data_raw.loc[data_raw[\"Func.refGene\"].str.contains(\"exonic|splicing\")]\n",
    "    # 去除synonymous SNV\n",
    "    data_sub = data_sub.loc[data_raw[\"ExonicFunc.refGene\"] != \"synonymous SNV\"]\n",
    "    # 只保留给定的基因列表的结果\n",
    "    data_sub = data_sub[data_sub[\"Gene.refGene\"].isin(gene_list)]\n",
    "    # 对于有CLINVAR有注释的结果的数据进行打分 (1)\n",
    "    data_sub['CLINSIG'] = data_sub['CLINSIG'].fillna(\".\")\n",
    "    data_sub.loc[data_sub['CLINSIG'].str.contains('Likely pathogenic|Pathogenic|drug response'), 'rating'] = 1\n",
    "    # 进行变异注释\n",
    "    data_sub[\"mutation_info\"] = data_sub.apply(get_mutation_info_varscan, axis=1)\n",
    "    data_sub[[\"tumorAF\", \"normalAF\", \"tumorDepth\", \"normalDepth\", \"tumorRefReads\", \"normalRefReads\", \"tumorMutReads\", \"normalMutReads\", \"tumorGenotype\", \"normalGenotype\"]] = pd.DataFrame(data_sub[\"mutation_info\"].tolist(), index=data_sub.index)\n",
    "    output_file_path = file_path.replace(\".csv.hg19_multianno.csv\", \"_processed.csv\")\n",
    "    # 直接输出大于5%的变异(rating=1)，对小于0.2%的进行打分(rating=0.5)\n",
    "    data_sub['tumorAF'] = data_sub['tumorAF'].fillna('-1')\n",
    "    data_sub.loc[data_sub['tumorAF'] > 5, 'rating'] = 1\n",
    "    data_sub.loc[data_sub['tumorAF'] < 0.2, 'rating'] = 0.5\n",
    "    # 对于变异reads数小于4的打分(rating=0.5)\n",
    "    data_sub['tumorMutReads'] = data_sub['tumorMutReads'].fillna('-1')\n",
    "    data_sub.loc[data_sub['tumorMutReads'] < 4, 'rating'] = 0.5\n",
    "    # 输出结果\n",
    "    data_sub.to_csv(output_file_path, index=False)\n",
    "    print(f\"Processed and saved: {output_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7319db7a-5992-4a48-85e2-16504361ee0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AKT1', 'ALK', 'BRAF', 'DDR2', 'EGFR', 'ERBB2', 'KRAS', 'MAP2K1', 'MET', 'NRAS', 'NTRK1', 'NTRK2', 'NTRK3', 'PIK3CA', 'PTEN', 'RET', 'ROS1', 'TP53']\n"
     ]
    }
   ],
   "source": [
    "# 读取18 gene list\n",
    "with open(\"/dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/filterMerge/18gene.txt\", \"r\") as f:\n",
    "    gene_list = [line.strip() for line in f.readlines() if line.strip()]\n",
    "\n",
    "print(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7574e2c6-77ab-4e92-ab88-44e4a2584747",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed and saved: /dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/Lib55036BT/ANNOVAR/Lib55036BT.Varscan_processed.csv\n",
      "Processed and saved: /dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/Lib55038BT/ANNOVAR/Lib55038BT.Varscan_processed.csv\n",
      "Processed and saved: /dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/Lib55040BT/ANNOVAR/Lib55040BT.Varscan_processed.csv\n",
      "Processed and saved: /dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/Lib55039BT/ANNOVAR/Lib55039BT.Varscan_processed.csv\n",
      "Processed and saved: /dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/202314/ANNOVAR/202314.Varscan_processed.csv\n",
      "Processed and saved: /dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/202315/ANNOVAR/202315.Varscan_processed.csv\n",
      "Processed and saved: /dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/202311/ANNOVAR/202311.Varscan_processed.csv\n",
      "Processed and saved: /dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/202312/ANNOVAR/202312.Varscan_processed.csv\n",
      "Processed and saved: /dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/Lib55037BT/ANNOVAR/Lib55037BT.Varscan_processed.csv\n",
      "Processed and saved: /dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA/202313/ANNOVAR/202313.Varscan_processed.csv\n"
     ]
    }
   ],
   "source": [
    "root_dir = \"/dssg/home/acct-medkwf/medkwf4/results/ctDNA_EQA\"\n",
    "for current_dir, subdirs, files in os.walk(root_dir):\n",
    "    target_files = [file for file in files if file.endswith(\"Varscan.csv.hg19_multianno.csv\")]\n",
    "    for file in target_files:\n",
    "        file_path = os.path.join(current_dir, file)\n",
    "        process_file(file_path, gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "9466a6ea-93b7-4720-a359-82a2f79e5809",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>CLINSIG</th>\n",
       "      <th>tumorAF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>NRAS</td>\n",
       "      <td>Likely pathogenic|Likely pathogenic|Likely pat...</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>DDR2</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>0.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>ALK</td>\n",
       "      <td>Likely pathogenic|not provided</td>\n",
       "      <td>0.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>Likely pathogenic|Likely pathogenic|Likely pat...</td>\n",
       "      <td>0.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>drug response|Likely pathogenic|Pathogenic|Lik...</td>\n",
       "      <td>2.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>Pathogenic|Likely pathogenic</td>\n",
       "      <td>0.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>Pathogenic|Pathogenic|not provided|Pathogenic|...</td>\n",
       "      <td>0.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Likely pathogenic|Likely pathogenic|Likely pat...</td>\n",
       "      <td>0.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>551</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Pathogenic|Pathogenic|Likely pathogenic|Likely...</td>\n",
       "      <td>9.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Likely pathogenic|Likely pathogenic|Likely pat...</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>554</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>575</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Likely pathogenic|Likely pathogenic|Likely pat...</td>\n",
       "      <td>1.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>Likely pathogenic|Likely pathogenic|Likely pat...</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ExonicFunc.refGene Gene.refGene  \\\n",
       "3    nonsynonymous SNV         NRAS   \n",
       "87   nonsynonymous SNV         DDR2   \n",
       "102  nonsynonymous SNV          ALK   \n",
       "159  nonsynonymous SNV       PIK3CA   \n",
       "266  nonsynonymous SNV         EGFR   \n",
       "450  nonsynonymous SNV         PTEN   \n",
       "467  nonsynonymous SNV         KRAS   \n",
       "545  nonsynonymous SNV         TP53   \n",
       "549  nonsynonymous SNV         TP53   \n",
       "551  nonsynonymous SNV         TP53   \n",
       "552  nonsynonymous SNV         TP53   \n",
       "554  nonsynonymous SNV         TP53   \n",
       "575  nonsynonymous SNV        ERBB2   \n",
       "614  nonsynonymous SNV        ERBB2   \n",
       "\n",
       "                                               CLINSIG  tumorAF  \n",
       "3    Likely pathogenic|Likely pathogenic|Likely pat...     0.28  \n",
       "87                                          Pathogenic     0.22  \n",
       "102                     Likely pathogenic|not provided     0.15  \n",
       "159  Likely pathogenic|Likely pathogenic|Likely pat...     0.23  \n",
       "266  drug response|Likely pathogenic|Pathogenic|Lik...     2.69  \n",
       "450                       Pathogenic|Likely pathogenic     0.29  \n",
       "467  Pathogenic|Pathogenic|not provided|Pathogenic|...     0.27  \n",
       "545                                  Likely pathogenic     0.18  \n",
       "549  Likely pathogenic|Likely pathogenic|Likely pat...     0.17  \n",
       "551  Pathogenic|Pathogenic|Likely pathogenic|Likely...     9.25  \n",
       "552  Likely pathogenic|Likely pathogenic|Likely pat...     0.18  \n",
       "554                                  Likely pathogenic     0.20  \n",
       "575  Likely pathogenic|Likely pathogenic|Likely pat...     1.86  \n",
       "614  Likely pathogenic|Likely pathogenic|Likely pat...     0.28  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
